src/core/scope.ml

```diff
@@ -408,45 +427,63 @@ let scope_term : expo -> sig_state -> env -> p_term -> term =
-    | P_Impl(a,b)        -> patt_vars (patt_vars acc a) b
-    | P_Abst(xs,t)       -> patt_vars (arg_patt_vars acc xs) t
-    | P_Prod(xs,b)       -> patt_vars (arg_patt_vars acc xs) b
-    | P_LLet(_,xs,a,t,u) ->
-        let pvs = patt_vars (patt_vars (arg_patt_vars acc xs) t) u in
+    | P_Meta(_,None)     -> acc
+    | P_Meta(_,Some ts)  -> Array.fold_left (patt_vars k) acc ts
+    | P_Patt(id,ts)      -> add_patt k acc id ts
+    | P_Appl(t,u)        -> patt_vars k (patt_vars k acc t) u
+    | P_Impl(a,b)        -> patt_vars k (patt_vars k acc a) b
+    | P_Abst(args,t)     -> patt_vars k (patt_vars_args k acc args) t
+    | P_Prod(args,b)     -> patt_vars k (patt_vars_args k acc args) b
+    | P_LLet(_,args,a,t,u) ->
+        let pvs = patt_vars k (patt_vars k (patt_vars_args k acc args) t) u in
         begin
           match a with
           | None    -> pvs
-          | Some(a) -> patt_vars pvs a
+          | Some(a) -> patt_vars k pvs a
         end
     | P_NLit(_)          -> acc
-    | P_UnaO(_,t)        -> patt_vars acc t
-    | P_BinO(t,_,u)      -> patt_vars (patt_vars acc t) u
-    | P_Wrap(t)          -> patt_vars acc t
-    | P_Expl(t)          -> patt_vars acc t
-  and add_patt ((pvs, nl) as acc) id ts =
+    | P_UnaO(_,t)        -> patt_vars k acc t
+    | P_BinO(t,_,u)      -> patt_vars k (patt_vars k acc t) u
+    | P_Wrap(t)          -> patt_vars k acc t
+    | P_Expl(t)          -> patt_vars k acc t
+  and add_patt k ((pvs, nl) as acc) id ts =
+    let acc, arity =
+      match ts with
+      | None -> (acc, k)
+      | Some(ts) -> (Array.fold_left (patt_vars k) acc ts, Array.length ts)
+    in
     match id with
     | None     -> acc
     | Some(id) ->
-    try
-      if List.assoc id.elt pvs <> Array.length ts then
-        fatal id.pos "Arity mismatch for pattern variable [%s]." id.elt;
-      if List.mem id.elt nl then acc else (pvs, id.elt :: nl)
-    with Not_found -> ((id.elt, Array.length ts) :: pvs, nl)
-  and arg_patt_vars acc xs =
-    match xs with
-    | []                    -> acc
-    | (_, None   , _) :: xs -> arg_patt_vars acc xs
-    | (_, Some(a), _) :: xs -> arg_patt_vars (patt_vars acc a) xs
+        begin
+          try
+            if List.assoc id.elt pvs <> arity then
+              fatal id.pos "Arity mismatch for pattern variable [%s]." id.elt;
+            if List.mem id.elt nl then acc else (pvs, id.elt :: nl)
+          with Not_found -> ((id.elt, arity) :: pvs, nl)
+        end
+  and patt_vars_args k acc args =
+    match args with
+    | []              -> acc
+    | (ids,a,_)::args ->
+        let acc = match a with None -> acc | Some a -> patt_vars k acc a in
+        (* We compute the number of named variables. *)
+        let rec nb_named_vars ids =
+          match ids with
+          | [] -> 0
+          | None :: ids -> nb_named_vars ids
+          | Some _ :: ids -> 1 + nb_named_vars ids
+        in
+        patt_vars_args (k + nb_named_vars ids) acc args
```

Comment:
Can you explain briefly what you did here?

Reply:
The arity of a generated pattern variable (for replacing some `_`) depends on the number of named variables. If you have `\x y,_` you get `_` replaced by `$M[x;y]`, and if you have `\x _,_` you get `_` replaced by `$M[x]`. But I am going to revert this change finally. See https://github.com/Deducteam/lambdapi/pull/412#issuecomment-638232127.